src/rust/engine/src/externs/options.rs

```diff
@@ -146,11 +161,121 @@ impl PyConfigSource {
+
+// The derivation of the option value, as a vec of (value, vec of source ranks) tuples.
+// A scalar value will always have a single source. A list/dict value may have elements
+// appended across multiple sources.
+// In ascending rank order, so the last value is the final value of the option.
+type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
+
+// A tuple (final value, rank of final value, optional derivation of value).
+//
+// Note: The final value and its rank could be computed from the derivation (see above),
+// but the full derivation is not itself computed in normal usage.
+// We could get rid of this tuple type by representing the final value and its rank as
+// a singleton derivation (in the case where we don't otherwise need the full derivation).
+// But that would allocate two unnecessary Vecs for every option.
+type OptionValue<T> = (Option<T>, isize, Option<OptionValueDerivation<T>>);
+
+// Condense list value derivation across sources, so that it reflects merges vs. replacements
+// in a useful way. E.g., if we merge [a, b] and [c], and then replace it with [d, e],
+// the derivation will show:
+//   - [d, e] (from command-line flag)
+//   - [a, b, c] (from env var, from config)
+fn condense_list_value_derivation<T: PartialEq>(
+    derivation: Vec<(Source, Vec<ListEdit<T>>)>,
+) -> OptionValueDerivation<Vec<T>> {
+    let mut ret: OptionValueDerivation<Vec<T>> = vec![];
+    let mut cur_group = vec![];
+    let mut cur_ranks = vec![];
+
+    // In this case, for simplicity, we always use the "inefficient" O(M*N) remover,
+    // even for hashable values. This is very unlikely to have a noticeable performance impact
+    // in practice. And if it does, it would only be when we generate option value derivation
+    // for help display, and not in regular usage.
+    // See comments on OptionParser::parse_list_hashable() for context.
+    fn remover<T: PartialEq>(list: &mut Vec<T>, to_remove: &[T]) {
+        list.retain(|item| !to_remove.contains(item));
+    }
+
+    for (source, list_edits) in derivation.into_iter() {
+        for list_edit in list_edits {
+            if list_edit.action == ListEditAction::Replace {
+                if !cur_group.is_empty() {
+                    ret.push((
+                        apply_list_edits::<T>(remover, cur_group.into_iter()),
+                        cur_ranks,
+                    ));
+                }
+                cur_group = vec![];
+                cur_ranks = vec![];
+            }
+            cur_group.push(list_edit);
+            cur_ranks.push(source.rank() as isize);
+        }
+    }
+    if !cur_group.is_empty() {
+        ret.push((
+            apply_list_edits::<T>(remover, cur_group.into_iter()),
+            cur_ranks,
+        ));
+    }
+
+    ret
+}
 
-fn to_py<T>(res: Result<OptionalOptionValue<T>, String>) -> PyResult<RankedVal<Option<T>>> {
+// Condense dict value derivation across sources,  so that it reflects merges vs. replacements
```

Comment:
This is similar logic to the list version above, but unlike in Python where dynamic named dispatch covers up a multitude of sins, it was ugly as hell to try and unify these. The impedance mismatch is too significant. 

So I figured it was better to have some near-repetition.